transform_hoist_assign.ml

```diff
@@ -0,0 +1,55 @@
+open Instr
+
+(* Hoisting assignments "x <- exp" as far up the callgraph as possible.
+ *
+ * Since we are not in SSA moving assignments is only possible (without further
+ * analysis) if the assignments dominates all its uses. Otherwise we might
+ * accidentally capture uses on some traces.
+ *
+ * The condition for a move to be valid is that the move target dominates the
+ * move origin (ie. we are moving upwards) and is dominated by all reaching
+ * defs (ie. we are not moving above our dependencies).
+ *
+ * We only look at our own use-def chain. Thus the transformation renames the
+ * variable to avoid overriding unrelated uses of the same name.
+ *)
+let hoist_assignment ((formals, instrs) as inp : analysis_input) : analysis_input option =
+  let open Analysis in
+  let reaching = reaching inp in
+  let uses = used inp in
+  let dominates = dominates inp in
+  let dominates_all_uses pc =
+    PcSet.for_all (fun use -> dominates pc use) (uses pc) in
+  let aliased = aliased inp in
+  let aliased var pc = VarSet.mem var (aliased pc) in
+  let rec find_possible_move pc =
+    if pc = Array.length instrs then None
+    else
+      let pc' = pc + 1 in
+      match[@warning "-4"] instrs.(pc) with
+      | Assign (x, exp) ->
+        if not (dominates_all_uses pc) || (aliased x pc) then find_possible_move pc'
```

Comment:
I don't have the operator precedences memorized, so I'm really paranoid about conditions like this (especially when negation is thrown in).